Pull request 6

vitsalis opened the pull request.

pkakelas committed:
renamed country flags

pkakelas committed:
renamed country flags

pkakelas committed:
Merge branch 'master' of https://github.com/VitSalis/endofcodes

pkakelas committed:
Configured country flags in user profile

pkakelas committed:
Merge branch 'master' of https://github.com/VitSalis/endofcodes

pkakelas committed:
Modified username validation rules

pkakelas committed:
Added flags path in configuratin file

Commit:
style

pkakelas committed:
fixed conflicts with vitsalis

pkakelas committed:
Renamed flags to lower case

pkakelas committed:
Strlower to country shortname

pkakelas committed:
Merge branch 'master' of https://github.com/VitSalis/endofcodes

pkakelas committed:
fixed merged stuff with Dionyziz

pkakelas committed:
Merge branch 'master' of https://github.com/dionyziz/endofcodes

vitsalis committed:
Add migration script for grader.

-Create table games.
-Create table creatures.
-Create table gameusers.
-Create table roundcreatures.
Check the changes in db_schema for information about the columns.

vitsalis committed:
Merge branch 'master' of https://github.com/pkakelas/endofcodes

vitsalis committed:
Fix style and merge from pkakelas

vitsalis committed:
Fix a stupid mistake

vitsalis committed:
Finished some corrections on db_schema

pkakelas committed:
Renamed some countries that were not in the country array

pkakelas committed:
Renamed some countries that were not in the country array

pkakelas committed:
Added countries in counties_array

pkakelas committed:
Created a counter in country population that counts all the elements of the array

pkakelas committed:
Deleted unescessary function that rejects spaces in username validation

vitsalis committed:
Remove auto_increment from creature.id

vitsalis committed:
Fix some database problems.

-Change the primary key of creatures table.
-Remove the unique key of roundcreatures table.
-Add a primary key in the roundcreatures table.

vitsalis committed:
Fix db_schema

vitsalis committed:
Fix my stupid mistakes

vitsalis committed:
Fix sql schema

vitsalis committed:
Change the names of the primary keys

vitsalis committed:
Accidentaly pushed a wrong change

dionyziz commented on models/user.php:
```diff
@@ -45,7 +45,7 @@ protected function validate() {
                 throw new ModelValidationException( 'username_empty' );
             }
-            if ( strpos( $this->username, ' ' ) || preg_match( '#[^a-zA-Z0-9]#', $this->username ) ) {
+            if ( strpos( $this->username, ' ' ) || preg_match( '#[^a-zA-Z0-9._]#', $this->username ) ) {
```
strpos must be checked against boolean false using ===, not like this; suggest moving space into regexp

pkakelas committed:
Merge branch 'master' of https://github.com/VitSalis/endofcodes

vitsalis committed:
population script did not catch exception

pkakelas committed:
Fixed conflicts with Vit

pkakelas committed:
Deleted unnecessary flag

vitsalis committed:
Do some changes in the db schema.

-Remove the unique key from table gameusers.
-Add a primary key to the table gameusers.
-Change type of roundcreatures.desire from varchar to enum.
-Add the changes to the db_schema.sql

pkakelas committed:
Merge branch 'master' of https://github.com/VitSalis/endofcodes

pkakelas committed:
Created a formtoken creation and validation functions

pkakelas committed:
Added a hidden input in register page that will have as a value a random token

pkakelas committed:
Forgotten semicolon

pkakelas committed:
Added a hidden input in session start page that will have as a value a random token

pkakelas committed:
Added csrf protection in register form

pkakelas committed:
Added csrf protection in login form

pkakelas committed:
Added hidden inputs user update forms

pkakelas committed:
Added csrf protection in the three update forms

pkakelas committed:
Fixed all Vitalis' corrections

dionyziz committed:
Just checking if I can upload from my new operating system

vitsalis committed:
Remove auto increment from db_schema

pkakelas committed:
Added csrf protection to logout form (Also changed privileges in my system)

pkakelas committed:
Closed a tag

pkakelas committed:
Changed formtoken encryption to openssl_random_pseudo_bytes

dionyziz commented on database/db_schema.sql:
```diff
@@ -31,7 +31,46 @@ CREATE TABLE IF NOT EXISTS `countries` (
+
+CREATE TABLE IF NOT EXISTS `games` (
+  `id` int(11) NOT NULL AUTO_INCREMENT,
+  `created` date NOT NULL,
```
DATETIME

vitsalis replied:
You said it's date because we have a game per day
On Dec 31, 2013 11:57 PM, "Dionysis Zindros" notifications@github.com
wrote:

> In database/db_schema.sql:
> 
> > +CREATE TABLE IF NOT EXISTS `creatures` (
> > -  `id` int(11) NOT NULL,
> > -  `gameid` int(11) NOT NULL,
> > -  `userid` int(11) NOT NULL,
> > -  PRIMARY KEY (`gameid`,`userid`,`id`)
> >   +) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;
> >   +
> >   +-- --------------------------------------------------------
> >   +
> >   +--
> >   +-- Table structure for table `games`
> >   +--
> >   +
> >   +CREATE TABLE IF NOT EXISTS `games` (
> > -  `id` int(11) NOT NULL AUTO_INCREMENT,
> > -  `created` date NOT NULL,
> 
> DATETIME
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/dionyziz/endofcodes/pull/8/files#r8603930
> .

dionyziz replied:
We might have more games in a single day later. Furthermore, it's good for logistic purposes to also record the time of the game, even if it happens once a day. We can display only the day to the users if we wish.

dionyziz commented on database/db_schema.sql:
```diff
@@ -42,9 +81,26 @@ CREATE TABLE IF NOT EXISTS `countries` (
+  `roundid` int(11) NOT NULL,
+  `gameid` int(11) NOT NULL,
+  `creatureid` int(11) NOT NULL,
+  `desire` enum('NORTH','WEST','EAST','SOUTH','NONE') COLLATE utf8_unicode_ci NOT NULL,
```
Would prefer to order this as per CSS's order: NONE, NORTH, EAST, SOUTH, WEST

dionyziz commented on views/user/view.php:
```diff
@@ -6,6 +6,10 @@
 
+<p><img src="<?php
+            echo $user->country->flag;
+        ?>" alt="Country flag" width="100" height="100" /></p>
```
Use country name in alt.

vitsalis committed:
Do some changes in the database.

-Change roundcreatures.desire enum type to ('NONE', 'NORTH', 'EAST', 'SOUTH', 'WEST')
-Change games.created data type to datetime type

vitsalis committed:
Change alt of country's flag image

pkakelas committed:
Merge branch 'master' of https://github.com/VitSalis/endofcodes

vitsalis committed:
Fix bug and add a test version of ControllerBase

vitsalis committed:
Fix a bug.

When a user that wasn't loggged in visited a user's page
there was an error. Fixed that.

vitsalis committed:
Move dispatcher logic from index.php to ControllerBase; change static controller method calls to non-static

vitsalis committed:
Refactor dispatching to achieve DRY in protection from CSRF forgery

vitsalis committed:
Now all the forms are created with the help of a class

dionyziz commented:
Please undo the unnecessary chmod changes.

dionyziz commented on controllers/dashboard.php:
```diff
@@ -1,9 +1,12 @@
                 $username = $_SESSION[ 'user' ][ 'username' ];
             }
+            include_once 'models/formtoken.php';
+            $token = FormToken::create();
```
DRY. This should be the responsibility of the Form class.

dionyziz commented on helpers/html.php:
```diff
@@ -1,4 +1,96 @@
+            $this->method = $method;
+        }
+
+        public static function produceError( $error_msg ) {
```
Why is this "produce" and the one below "create"?

dionyziz commented on helpers/html.php:
```diff
@@ -1,4 +1,96 @@
+
+        public static function produceError( $error_msg ) {
+            ?><p class="error"><?php
+                    echo $error_msg;
```
indentation

dionyziz commented on helpers/html.php:
```diff
@@ -1,4 +1,96 @@
+
+        public static function createInput( $type = 'text', $name = '', $id = '', $value = '' ) {
+            ?><p><input type="<?php
+                echo $type;
```
whitelist types

dionyziz commented on helpers/html.php:
```diff
@@ -1,4 +1,96 @@
+                }
+                if ( !empty( $value ) ) {
+                    ?>value="<?php
+                        echo $value;
```
XSS

dionyziz commented on helpers/html.php:
```diff
@@ -1,4 +1,96 @@
+                ?><option <?php
+                    if ( isset( $option[ 'value' ] ) ) {
+                        ?>value="<?php
+                            echo $option[ 'value' ];
```
XSS

dionyziz commented on helpers/html.php:
```diff
@@ -1,4 +1,96 @@
+            ?><label for="<?php
+                echo $for;
+            ?>"><?php
+                echo $text;
```
XSS

dionyziz commented on helpers/html.php:
```diff
@@ -1,4 +1,96 @@
+                    echo $this->resource;
+                ?>&amp;method=<?php
+                    echo $this->method;
+                ?>" method="POST"><?php
```
- Lower-case "post".
- Some forms may need to be "get".
- This doesn't work for file uploads; detect if there are file inputs and send correct enctype

dionyziz commented on helpers/html.php:
```diff
@@ -1,4 +1,96 @@
+        private $resource;
+        private $method;
+        public $id;
+        public $token;
```
protect \$token and own it; the caller should not concern themselves with this

dionyziz commented on helpers/html.php:
```diff
@@ -1,4 +1,96 @@
 <?php
+    class Form {
+        private $resource;
+        private $method;
```
make these protected for extensibility

dionyziz commented on views/session/create.php:
```diff
@@ -2,35 +2,33 @@
+            $username_wrong, $password_wrong ) {
         if ( isset( $username_empty ) ) {
-            ?><p class="error">Please type a username.</p><?php
+            Form::produceError( "Please type a username." );
```
Make these non-static method calls. Form needs to maintain state.

vitsalis replied:
how am I going to do that?

dionyziz commented on views/session/create.php:
```diff
@@ -2,35 +2,33 @@
-            }
-        ?>
-        <p><input type="password" name="password" id="password" /></p>
-        <p><a href="">Forgot password?</a></p>
```
Where did this go?

dionyziz commented on views/user/create.php:
```diff
@@ -20,128 +20,95 @@
-</form>
+            $email_value = "";
+        }
+        Form::createInput( 'text', 'email', 'email', htmlspecialchars( $email_value ) );
```
Move htmlspecialchars to form; pass actual value here.

dionyziz commented on views/user/create.php:
```diff
@@ -20,128 +20,95 @@
+        Form::createInput( 'text', 'email', 'email', htmlspecialchars( $email_value ) );
+        Form::createLabel( 'dob', 'Date of birth' );
+        $days_select_array = array();
+        $days_select_array[] = array( 'content' => 'Select Day' );
```
Combine these two lines.

dionyziz commented on views/user/create.php:
```diff
@@ -20,128 +20,95 @@
+        }
+        Form::createSelect( 'day', 'dob', $days_select_array );
+        $months_select_array = array();
+        $months_select_array[] = array( 'content' => 'Select Month' );
```
Combine

dionyziz commented on views/user/create.php:
```diff
@@ -20,128 +20,95 @@
+        }
+        Form::createSelect( 'month', '', $months_select_array );
+        $years_select_array = array();
+        $years_select_array[] = array( 'content' => 'Select Year' );
```
Combine

dionyziz commented on views/user/create.php:
```diff
@@ -20,128 +20,95 @@
+        $years_select_array = array();
+        $years_select_array[] = array( 'content' => 'Select Year' );
+        $current_year = date( 'Y' );
+        for ( $i = $current_year - 8; $i >= $current_year - 100; --$i ) {
```
min age / max age should be configurable constants

dionyziz commented on views/user/create.php:
```diff
@@ -20,128 +20,95 @@
+        }
+        Form::createSelect( 'year', '', $years_select_array );
+        $countries_select_array = array();
+        $countries_select_array[] = array( 'content' => 'Select Country' );
```
combine

dionyziz commented on views/user/create.php:
```diff
@@ -20,128 +20,95 @@
+        $countries_select_array = array();
+        $countries_select_array[] = array( 'content' => 'Select Country' );
+        foreach ( $countries as $key => $country ) {
+            $countries_select_array[] = array( 'value' => $key + 1, 'content' => $country[ 'name' ] );
```
Why +1?

vitsalis replied:
because at database the countries id starts from 1 but here they keys start from 0

dionyziz commented on views/user/update.php:
```diff
@@ -1,72 +1,61 @@
-    <p><input type="file" name="image" id="image" /></p>
-    <input type="submit" value="Upload" />
-</form>
+        Form::createInput( 'file', 'image', 'image' );
```
Did you test this?

vitsalis committed:
Create token in the Form class

vitsalis committed:
Rename produceError to createError

vitsalis committed:
Fix wrong number of tabs in a line

vitsalis committed:
Add a whitelist of types

vitsalis committed:
Fix a bug and do some refactoring.

-Now the forms can have 'get' methods too.
-Image upload works.
-Fix some XSS problems.

vitsalis committed:
Do resource and method protected instead of private

vitsalis committed:
Combine declaration of arrays, add max and min age to config

vitsalis committed:
Undo permissions bullshit

dionyziz commented on helpers/html.php:
```diff
@@ -1,4 +1,147 @@
+                'url',
+                'week'
+            );
+            foreach ( $valid_types as $valid_type ) {
```
array_flip() and isset()

vitsalis committed:
Remove unnesecary token creation

dionyziz commented on views/user/update.php:
```diff
@@ -1,72 +1,59 @@
-    <label for="image">Upload an avatar</label>
-    <?php
+    $form = new Form( 'image', 'create' );
+    $form->hasFile = true;
```
Don't set this here. Let the form take care of it when you createInput( 'file' ... );

dionyziz commented on helpers/html.php:
```diff
@@ -1,4 +1,147 @@
+        protected $method;
+        public $id;
+        public $formMethod;
+        public $hasFile = false;
```
protected

dionyziz commented on views/session/create.php:
```diff
@@ -2,36 +2,36 @@
+<div id="login"><?php 
+    $form = new Form( 'session', 'create' );  
+    $form->id = 'login-form';
+    $form->formMethod = 'post';
```
Let the form handle this based on method passed to constructor. For idempotent methods, use GET. For non-idempotent methods, use POST.

vitsalis committed:
Remove formtoken from controllers

vitsalis committed:
make static functions non-static and do some refactoring

vitsalis committed:
Add array_flip instead of searching all of the array

vitsalis committed:
Remove useless foreach

vitsalis committed:
Add HTMLFormInvalidException and remove a static call to a non-static function

vitsalis committed:
Include helpers/html.php on dependencies, fix grammatical error, make getRESTMethodIdempotence static

vitsalis committed:
DRY. Use the Form class in the dispatcher

dionyziz commented on helpers/html.php:
```diff
@@ -1,4 +1,182 @@
+                'week'
+            );
+            $valid_types = array_flip( $valid_types );
+            if ( isset( $valid_types[ $type ] ) ) {
```
`return isset...`
